storm/src/main/storm/mesos/schedulers/DefaultScheduler.java

```diff
@@ -201,28 +214,55 @@ public void prepare(Map conf) {
-      // Un-assign them
-      int slotsFreed = schedulerAssignment.getSlots().size();
-      cluster.freeSlots(schedulerAssignment.getSlots());
-      log.info("executorsPerWorkerList - slotsAvailable: {}, slotsAssigned: {}, slotsFreed: {}", slotsAvailable, slotsAssigned, slotsFreed);
-      executors = cluster.getUnassignedExecutors(topologyDetails);
-      slotsAvailable += slotsAssigned;
-    } else if (executors.isEmpty()) {
-      log.info("There are currently no unassigned executors. Not going to redistribute work because slotsAvailable is {} and slotsAssigned is {}", slotsAvailable, slotsAssigned);
+    int slotsToUse = 0;
+
+    // If there are not any unassigned executors, we need to re-distribute all currently assigned executors across workers
+    if (executors.isEmpty()) {
+      if (slotsAssigned < slotsAvailable) {
+        log.info("All executors are already assigned, but only onto {} slots. Redistributing all assigned executors to new set of {} slots.", slotsAssigned, slotsAvailable);
+        SchedulerAssignment schedulerAssignment = cluster.getAssignmentById(topologyDetails.getId());
+        // Un-assign them
+        int slotsFreed = schedulerAssignment.getSlots().size();
+        cluster.freeSlots(schedulerAssignment.getSlots());
+        log.info("executorsPerWorkerList - slotsAvailable: {}, slotsAssigned: {}, slotsFreed: {}", slotsAvailable, slotsAssigned, slotsFreed);
+        executors = cluster.getUnassignedExecutors(topologyDetails);
+        // Need to instead wipe *everything* and just use the slotsAvailable
+        slotsToUse = slotsAvailable;
+      } else {
+        log.info("All executors are already assigned. Not going to redistribute work because slotsToUse is {} and slotsAssigned is {}", slotsAvailable, slotsAssigned);
+        return null;
+      }
+    } else {
+      /*
+       * Spread the unassigned executors onto however many available slots we can possibly use.
+       * i.e., there might be more than we need.
+       *
+       * Note that this logic can lead to an imbalance of executors/worker between various workers.
+       *
+       * We propose to avoid such problems by having an option (perhaps on by default) which only will
+       * ever schedule onto the exact requested number of workers.
+       * See https://github.com/mesos/storm/issues/158
+       * For now we just issue a warning when we detect such a situation.
+       */
+      int slotsNeeded = slotsRequested - slotsAssigned;
+      slotsToUse = Math.min(slotsNeeded, slotsAvailable);
+      // Notably, if slotsAssigned was 0, then this would be a full rebalance onto less workers than requested,
+      // and hence wouldn't lead to an imbalance.
+      if (slotsToUse + slotsAssigned < slotsRequested && slotsAssigned != 0) {
+        log.warn("we are assigning storm executors for topologyId: {} onto less slots ({}) than requested ({}), " +
```

Comment:
Maybe this could be 
```
log.warn("Assigning storm executors for {} onto less slots ({}) than requested ({}), " +
         "this may lead to executor imbalance.", 
         topologyDetails.getId(), slotsToUse + slotsAssigned, slotsRequested);
```
since saying that the topologyId is the topologyId seems unnecessary and the explanation is more concise?